### PR TITLE
Check if "CMakeLists.txt" exists, currently or after renaming

### PR DESCRIPTION
**Pull request by elahehrashedi**
bugfix: https://github.com/microsoft/vscode-cmake-tools/issues/2986
Works for creating and deleting a CMakeLists.txt file

Added a "rename" situation as well.

### CODE REVIEW

**Review by nasinlock — APPROVED** [review]

**Review by nasinlock — APPROVED** [review]

File: src/cmakeProject.ts [hunk 1]
```diff
@@ -1522,6 +1522,7 @@ export class CMakeProject {
         if (isCmakeFile) {
             // CMakeLists.txt change event: its creation or deletion are relevant,
             // so update full/partial feature set view for this folder.
+            this.cmakeListsExists = await fs.exists(path.join(this._sourceDir, "CMakeLists.txt"));
```
Comment (bobbrow) [inline src/cmakeProject.ts:1525]: It doesn't seem like this should be necessary since whenever `this._sourceDir` is set, `this.cmakeListsExists` should also be set.

File: src/cmakeProject.ts [hunk 2]
```diff
@@ -2583,8 +2594,8 @@ export class CMakeProject {
         return expandStrings(this.workspaceContext.config.additionalKits, opts);
     }
 
-    async sendFileTypeTelemetry(textDocument: vscode.TextDocument): Promise<void> {
-        const filePath =  util.platformNormalizePath(textDocument.uri.fsPath);
+    async sendFileTypeTelemetry(textDocument: vscode.Uri): Promise<void> {
```
Comment (bobbrow) [inline src/cmakeProject.ts:2597]: This change seems unnecessary. Please revert.

File: src/cmakeProject.ts [hunk 3]
```diff
@@ -1493,8 +1493,8 @@ export class CMakeProject {
     }
 
     // Reconfigure if the saved file is a cmake file.
-    async doCMakeFileSaveReconfigure(textDocument: vscode.TextDocument) {
-        const filePath = util.platformNormalizePath(textDocument.uri.fsPath);
+    async doCMakeFileChangeReconfigure(textDocument: vscode.Uri) {
```
Comment (bobbrow) [inline src/cmakeProject.ts:1496]: Changing the type of the argument seems unncessary. Please revert to `vscode.TextDocument`.

File: src/extension.ts [hunk 4]
```diff
@@ -1425,7 +1425,7 @@ export class ExtensionManager implements vscode.Disposable {
             return false;
         }
         for (const project of projects) {
-            if (await project.hasCMakeLists()) {
+            if (project.hasCMakeLists()) {
```
Comment (bobbrow) [inline src/extension.ts:1428]: If the `await` is gone, consider removing the `async` specifier from this function.

Comment (bobbrow) [reply src/extension.ts:1428]: You could also use `return projects.some(p => p.hasCMakeLists());` instead of a for-loop here too.

File: src/projectController.ts [hunk 5]
```diff
@@ -117,7 +117,7 @@ export class ProjectController implements vscode.Disposable {
     async getNumOfValidProjects(): Promise<number> {
```
Comment (bobbrow) [inline src/projectController.ts:117]: It looks like this doesn't have to be `async` anymore.

File: src/projectController.ts [hunk 6]
```diff
@@ -441,11 +442,24 @@ export class ProjectController implements vscode.Disposable {
     }
 
     private async doSaveTextDocument(textDocument: vscode.TextDocument): Promise<void> {
+        await this.doCMakeFileChangeReconfigure(textDocument.uri);
+    }
+
+    private async onDidRenameFiles(renamedFileEvt: vscode.FileRenameEvent): Promise<void> {
```
Comment (bobbrow) [inline src/projectController.ts:448]: this doesn't seem related to `hasCMakeLists`?

File: src/util.ts [hunk 7]
```diff
@@ -864,9 +864,9 @@ export async function scheduleAsyncTask<T>(task: () => Promise<T>): Promise<T> {
     });
 }
 
-export function isFileInsideFolder(openEditor: vscode.TextDocument, folderPath: string): boolean {
+export function isFileInsideFolder(openEditor: vscode.Uri, folderPath: string): boolean {
```
Comment (bobbrow) [inline src/util.ts:867]: I don't think this needs to change. Please revert.

**Review by bobbrow — CHANGES_REQUESTED** [review]

File: src/util.ts (hunk 7 above)
Comment (bobbrow) [reply src/util.ts:867]: Oh, I see that the rename event is the cause for this signature change.  I think that should have been split out from this PR.

Comment (elahehrashedi) [reply src/util.ts:867]: The rename event only checks for existance of CMakeLists.txt. That's why I added it to this PR.

```ts
            if (filePath.endsWith("cmakelists.txt")) {
                await this.doCMakeFileChangeReconfigure(file.newUri);
            }
```

Comment (elahehrashedi) [reply src/util.ts:867]: I changed the title of the PR

File: src/cmakeProject.ts (hunk 3 above)
Comment (elahehrashedi) [reply src/cmakeProject.ts:1496]: This is related to checking if a renamed file is a "CMakeLists.txt" file. in "onDidRenameFiles" we had access to URI.

File: src/cmakeProject.ts (hunk 1 above)
Comment (elahehrashedi) [reply src/cmakeProject.ts:1525]: that's right!

File: src/cmakeProject.ts (hunk 2 above)
Comment (elahehrashedi) [reply src/cmakeProject.ts:2597]: This is related to checking if a renamed file is a "CMakeLists.txt" file. in "onDidRenameFiles" we had access to URI.

**Comment by elahehrashedi** [comment]
@bobbrow I applied some renaming.

**Review by bobbrow — APPROVED** [review]
